Rephrase (#3633)
ROCm-Developer-Tools/HIP: docs/how-to/stream_ordered_allocator.rst

> Using SOMA

Reviewer: Here's my comment for Line 6.
The topic should be titled:
Stream-Ordered Memory Allocator
This is a compound modifier so a hyphen should be used. 
It should be changed in other sections too

=====================================

You can allocate memory using ``hipMallocAsync()`` with stream-ordered
semantics. This restricts the asynchronous accesses to the memory to occur
between the stream executions of the allocation and deallocation. Accessing
memory outside this promised stream order can lead to undefined behavior such as

> use-before-allocation or use-after-free errors. The allocator might reallocate

Reviewer: ```suggestion
semantics. This restricts the asynchronous access to the memory between the stream executions of the allocation and deallocation. Accessing

> use-before-allocation or use-after-free errors. The allocator might reallocate
> 
> .. guarantee:
> 
> memory as long as the compliant memory accesses are guaranteed not to overlap
> temporally. ``hipFreeAsync()`` frees memory from the pool with stream-ordered
> semantics.

Reviewer: ```suggestion
use-before-allocation or use-after-free errors. The allocator might reallocate
memory if the compliant memory accesses won't overlap
temporally. ``hipFreeAsync()`` frees memory from the pool with stream-ordered
semantics.
```

The ``hipMallocAsync()`` function uses the current memory pool and also provides the opportunity to create and access different pools using ``hipMemPoolCreate()`` and ``hipMallocFromPoolAsync()`` functions respectively.

> .. always:
> 
> Unlike NVIDIA CUDA, where stream-ordered memory allocation can be implicit, in ROCm HIP, it's always explicit. This requires you to manage memory allocation for each stream in HIP while ensuring precise control over memory usage and synchronization.

Reviewer: ```suggestion
Unlike NVIDIA CUDA, where stream-ordered memory allocation can be implicit, ROCm HIP is always explicit. This requires managing memory allocation for each stream in HIP while ensuring precise control over memory usage and synchronization.
```

The ``hipMallocAsync()`` function uses the current memory pool and also provides the opportunity to create and access different pools using ``hipMemPoolCreate()`` and ``hipMallocFromPoolAsync()`` functions respectively.

.. always:

> Unlike NVIDIA CUDA, where stream-ordered memory allocation can be implicit, in AMD HIP, it's always explicit. This requires you to manage memory allocation for each stream in HIP while ensuring precise control over memory usage and synchronization.

Reviewer: ```suggestion
Unlike NVIDIA CUDA, where stream-ordered memory allocation can be implicit, in ROCm HIP, it's always explicit. This requires you to manage memory allocation for each stream in HIP while ensuring precise control over memory usage and synchronization.
```

Reviewer: It's never AMD HIP - it's ROCm HIP

Memory reuse policies
---------------------

> .. guarantee
> 
> The allocator might reallocate memory as long as the compliant memory accesses are guaranteed not to overlap temporally. To optimize the memory usage, disable or enable the following memory pool reuse policy attribute flags:

Reviewer: ```suggestion
The allocator might reallocate memory as long as the compliant memory accesses will not to overlap temporally. To optimize the memory usage, disable or enable the following memory pool reuse policy attribute flags:
```

Memory reuse policies
---------------------

.. guarantee

> The allocator might reallocate memory as long as the compliant memory accesses are guaranteed not to overlap temporally. To optimize the memory usage, disable or enable the following memory pool reuse policy attribute flags:

Reviewer: ```The allocator might reallocate memory as long as the compliant memory accesses don't overlap temporally. To optimize the memory usage, disable or enable the following memory pool reuse policy attribute flags:
```

Author: These comment seems to be broken

Device accessibility for multi-GPU support
------------------------------------------

> .. only:
> 
> Allocations are initially accessible only from the device where they reside.

Reviewer: ```suggestion
Allocations are initially accessible from the device where they reside.
```

Device accessibility for multi-GPU support
------------------------------------------

.. only:

> Allocations are initially accessible only from the device where they reside.

Reviewer: ```Allocations are initially accessible from the device where they reside.
```
